docs/proposals/config_data.md

```diff
@@ -0,0 +1,327 @@
+
+A new resource for ``ConfigData`` will be added to the API:
+
+```go
+	type ConfigData struct {
+		TypeMeta   `json:",inline"`
+		ObjectMeta `json:"metadata,omitempty"`
+
+		Data map[string]string `json:"data,omitempty"`
+	}
+
+	type ConfigDataList struct {
+		TypeMeta `json:",inline"`
+		ListMeta `json:"metadata,omitempty"`
+
+		Items []ConfigData `json:"items"`
+	}
+```
+
+A new API resource selector (ResourceSelector) will be added to select objects and fields from those objects
+for consuming API resources through the downward API.
+
+``Registry`` interface will be added to accompany the ``ConfigData`` resource.  
+``ConfigData`` information will be stored in ``etcd`` by default.
+
+### Volume Source
+
+A new ``ConfigDataVolumeSource`` type of volume source containing the ``ConfigData`` object will be added to the ``VolumeSource`` struct in the API:
+
+```go
+	type ConfigDataVolumeSource struct {
+		ConfigDataName string `json:"configDataName"`
+	}
+```
+
+**Note:** The update logic used in the downward API volume plug-in will be extracted and re-used in the volume plug-in for ``ConfigData``.
+
+## Creating ConfigData
+``ConfigData`` can be created in two ways explicitly and implicitly.
+
+### Backwards-Compatibility
+In order to smooth the transaction of component configuration from other means to ``ConfigData``
+and preserve backwards-compatibility, user will be able to configure components the old way
+and use an optional ``--write-config=<path>`` flag to create a ``ConfigData`` object containing
+the given configuration options to ``<path>``, which can be created by invoking ``kubectl create -f <path>``.
+
+For example, a ``kubelet`` will be configured as follows:
+```bash
+kubelet --address=0.0.0.0 --port=8888 --api-servers=10.240.13.14:8998,10.240.51.5:8998 --register-node=true --write-file=kubelet-config.yaml
+```
+
+The persisted object will have the following contents for the given configuration parameters:
+```json
+{
+    "apiVersion": "v1",
+    "kind": "ConfigData",
+    "data": {
+        "ADDRESS": "0.0.0.0",
+        "PORT": "8888",
+        "API_SERVERS": "10.240.13.14:8998, 10.240.51.5:8998",
+        "REGISTER_NODE": "true",
+    }
+}
+```
+
+An identically configured kubelet can be created as such from now on:
+```bash
+kubelet --config-data=kubelet-config.yaml
+```
+
+The command-line flags given to the ``kubelet`` will be transformed to ``Data`` fields
+inside a new ``ConfigData`` object for its configuration.
+
+**Note:** Tooling will be implemented to facilitate this functionality for each
+component that is going to migrate to ``ConfigData``.
+
+### Explicit Creation
+The following file ``pod-config-example.json`` contains data intended to be consumed as environment variables.
+```json
+{
+    "apiVersion": "v1",
+    "name": "pod-config-example",
+    "kind": "ConfigData",
+    "data": {
+        "server_address": "10.240.13.14:8998",
+        "db_address": "10.240.55.3",
+        "db_port": "3306"
+    }
+}
```

Comment:
Is there any way in this API to disable env variables? As someone actually working with an app, I'm finding them annoying to work around (the service environment variables in particular), so I'd prefer if there were fewer surprising env variables.